### ISSUE TITLE
Entity support via an interface needed

### ISSUE THREAD

**Issue by JC008** [rasmus/EventFlow#144]
@rasmus EventFlow needs an `IEntity` interface to map to the DDD concept of an Entity vs ValueObject. The interface could be simple and should not alter the current `IAggregateRoot<,>` approach except for __maybe__ exposing a `.Root` property of type `IEntity`.

```c#
public interface IEntity
{
    IIdentity Identity { get; }
}
```

Your planned domain example (Library) can be used to illustrate the issue and suggested change.

```c#
public class Book : IEntity
{
    public Isbn Isbn { get; } // <-- makes Book an Entity and not a Value Object per DDD conventions
    public Title Title { get; }
    public IIdentity Identity => Isbn;

    public Book(
        Isbn isbn,
        Title title)
    {
        Isbn = isbn;
        Title = title;
    }
}

public class Isbn : SingleValueObject<string>, IIdentity
{
    public Isbn(string value) : base(value)
    {
    }
}

public class Title : SingleValueObject<string>
{
    public Title(string value) : base(value)
    {
    }
}
```

The modified `Book` can be placed in a `<TopLevelNameSpace>/Entities` namespace instead of the  `<TopLevelNameSpace>/ValueObjects`.

The approach will then map to the DDD conventions of _ValueObject_ and _Entity_ without disruption in EventFlow implementations.

Your thoughts?

**Comment by rasmus** [comment]
How about creating these interfaces and classes to better support identities of specific types?

```csharp
public interface IEntity
{
  IIdentity GetIdentity();
}

public interface IEntity<TIdentity> : IEntity
  where TIdentity : IIdentity
{
  TIdentity Id { get; }
}

public abstract Entity<TIdentity> : IEntity<TIdentity>
  where TIdentity : IIdentity
{
  public TIdentity Id { get; }

  protected Entity(TIdentity id)
  {
    Id = id;
  }

  public IIdentity GetIdentity()
  {
    return Id;
  }
}
```

It would enable you to access `aBook.Id` with `Id` being of type `Isbn` instead of `IIdentity`?

I don't think I would add it to the `IAggregateRoot<>` has the specific type is put into e.g. `IEventStore.LoadAggregateAsync<,>` enabling you to put the `.Root` on your own interface/class.

If I understand it, then this wouldn't introduce any other code changes, but it would provide developers with a common `IEntity` interface that they could use?

**Comment by rasmus** [comment]
@JC008 Some like #145

**Comment by JC008** [comment]
@rasmus Correct - there are is no impact to any EventFlow code while it formalizes the concept of `Entity`. Your suggested interface and generic variant also makes the concept clear, usable and practical. Most importantly, it maps to the DDD concept and can optionally be introduced if needed in an implementation.

I would suggest adding a comment in the documentation regarding the missing `.Root` in `IAggregateRoot<.>` and putting it in your own interface/class to give guidance to developers to avoid lots of `Where is the .Root in the Aggregate?` questions.

**Comment by rasmus** [comment]
@JC008 But, wouldn't the class implementing `IAggregateRoot<>` be the root object itself negating the use of a `.Root` property? If not, then the `IAggregateRoot<>` implementation would not live up to its name and would basically be reduced to a big route for events appliers?

**Comment by JC008** [comment]
@rasmus My understanding is that an **Aggregate** may contain or use many entities but only one is the root which the outside world should be aware of. But, and **here** is where your **opinionated** implementation of EventFlow comes in, because EventFlow is an *Event Sourcing First* implementation, entities as a persistence management mechanism is not used and surfacing the `.Root` not possible as it does not exist - hence the use of `IAggregateRoot<,>`.

With that said and with the understanding that all domain interactions should be done via `ICommand` and `IQuery` by the consumer (e.g. programmer making use of an EventFlow based implementation), I find I need an `IEntity`, not from a persistence or model perspective, but from a data type perspective, that's why I used your Library example - it illustrated my problem.

At first, I struggled with the `IAggregateRoot<,>` concept and think some people will have an issue with the name - thinking it should be`IAggregate<,>`. If you look at your naming in the Library domain example,

```
c#  public class BookAggregate : AggregateRoot<BookAggregate, Isbn> ...
```

you will notice you created a `BookAggregate` - not a `BookAggregateRoot` - which is expected. The interface `IAggregateRoot<,>` is indeed correct and should remain so because of the **Event Sourcing** element, in effect making it a `root` at the interface in the architecture. Developers should make use of the [Service](https://lostechies.com/jimmybogard/2008/08/21/services-in-domain-driven-design/) concept in DDD if they don't have an explicit ES entry point implementable as `IAggregateRoot<,>` and should be **weened** of an *Entity First* approach.

**Comment by rasmus** [comment]
@JC008 So from what I understand from your comment, you basically would like to have a separate class for your DDD aggregate root being of e.g. type `Book` and implementing `IEntity`. You are limited by EventFlow as the `IAggregateRoot<>` root is tied up to the event sourcing part and you would not have a entity that implement both the EventFlow `IAggregateRoot<>` and the new `IEntity` interfaces. Is that correctly understood?

I have been think about removing the event sourcing part of the aggregate root and putting it into a new "event sourced" class instead. I did a small PoC on it in #118, but I think that attempt is flawed.

As an example, would the ideal solution be if you could have a the `IEventStore` work on a `IEventSourced<>` object that has a `IEventSource.Root` property of the type `IAggregateRoot<>` and `IAggregateRoot<>` implements `IEntity<>`? The event sourcing would have to be removed from the `IAggregateRoot<>` and moved into `IEventSourced`. The `IAggregateRoot<>` could be completely removed.

**Comment by JC008** [comment]
@rasmus I am happy with EventFlow as is - there are no limitations for me and no changes are needed except for a formal `IEntity<>`. 

My experience is that I keep creating `ValueObject(s)` that are really `Entity(ies)` because they have id's, same as your `Book` value object and therefor are `Entity(ies)`- hence the request. When I started using EventFlow I expected to see `Entity(ies)` but the **Event Sourcing** elements explained the lack of formal `Entity(ies)` support. 

I just reread the definitions from [DDD Reference](https://domainlanguage.com/ddd/reference/), [Fowler's DDD Aggregate](http://martinfowler.com/bliki/DDD_Aggregate.html), [cqrs.nu/Faq/aggregates](http://cqrs.nu/Faq/aggregates) and think we need to clarify the following:

1. A DDD aggregate is a larger unit of encapsulation than just a class
2. A DDD aggregate always has a root object
3. References to the aggregate from the outside world is only to the root object
4. To be referenced from the outside world the root object requires an identity
5. Because it has an identity it is an *entity* in DDD
6. The term *aggregate root* refers to an attribute of the *aggregate* not the **aggregate** itself
7. An *aggregate root* cannot exist in isolation in theory (see #8)
8. `IAggregateRoot<,>` is an **Event Sourcing** specific adaptation in EventFlow as I tried to suggest earlier

If EventFlow remains an *Event Sourcing First* system - no changes are needed, but if you want to enable existing entity modeled projects to phase into CQRS and ES  or have hybrid projects (a likely real world scenario) you will need to refactor as you suggested in your comment.

1. `IAggregateRoot<>` becomes `IAggegate<IEntity>`
2. Identity is provided by `IEntity.Id`
3. `IEntity` can optionally be `IEventSourced` or  use alternative providers e.g. `ISqlSourced`, `IRESTSourced` etc
4. Entity is now the `.Root` and can implement `IAggregateState`
5. `IAggregate` emits `IDomainEvent` and not `IAggregateEvent`
6. Aggregate coordinates behaviour between its contained objects (entities e.g. Invoice -> Line Items, Products, Discounts, Delivery Options etc etc)
7. `Entity` can be passed around as a reference
8. Extension methods can be added to an `IAggregate` implementation
9. Aggregate extension methods can pass calls to a resolved command bus e.g. `Invoice.Cancel()` on the `Entity` passes a `CancelInvoice(...)` command - looks like CRUD - behaves like CQRS
10. Can make EventFlow objects more accessible for consumers e.g. via collections
11. Hybrid solutions can use both event sourcing and non event sourced data

LUNCH!!!!!

**Comment by rasmus** [comment]
I totally forgot our discussion here, sorry.

- Aggregates could be modeled in EventFlow simply by using the `AggregateState<,,>` functionality recently added
- Personally I like the idea the entities are not directly liked to a aggregate root. It _could_ be, but it might also be some aggregate within the aggregate root
- In the shipping example I created for EventFlow I mapped entities from read models in query handlers. I think this provides a good abstraction to the entities as no memory reference is passed around to any aggregates, only entities
- As entities are passed around and not aggregate and entities are based on read models, the using them within an aggregate becomes something that's up to the individual developer and I don't think that EventFlow should impose concepts here
- As it is now, its fairly simple to model new aggregate roots as you only need one class and I would really like to keep it that way
- Extension methods are great to help developers to to the right think, and creating extension methods for your entities is a good idea, but I don't think I can generalize it. I maybe I could put it into the do and don't document for reference
- The case of having multiple event sources is interesting, but seems like a special case and I don't think I'll add support in EventFlow at the moment. If you have specific ideas on how to proceed, then create a PR or an issues with that
- I think that many applications have areas that doesn't use EventFlow or DDD, but I would like that EventFlow doesn't make any assumptions at all on how these parts are structured. That would make EventFlow too invasive

In short, I think with the addition of `IEntity` and `AggregateState<,,>` this issue can be closed. @JC008 do you agree or did I miss the point?

**Comment by JC008** [comment]
@rasmus I agree on all points and the current EventFlow implementation is spot on except for a small pedantic issue, `AggregateRoot<,,>` should be named `Aggregate<,,>` :wink:. Look at the classes below to see how I composed a **standardized** aggregate with an indirect `root` accessible via the aggregate's state. EventFlow is composable and flexible enough to handle variations in approach. We can close this one.

```csharp

public abstract class BaseAggregate<TAggregate, TAggregateState, TIdentity>
    : AggregateRoot<TAggregate, TIdentity>
    where TAggregate : AggregateRoot<TAggregate, TIdentity>
    where TAggregateState : class, IAggregateState<TAggregate>, IEventApplier<TAggregate, TIdentity>
    where TIdentity : class, IIdentity
{
    protected BaseAggregateRoot(TIdentity id)
        : base(id)
    {
        id.CheckForNull(nameof(id));

        State = Runtime.Resolver.Resolve<TAggregateState>(new TypedParameter(typeof (TIdentity), id));
        Register(State);
    }

    protected virtual void EmitError<TEvent>
        (TEvent aggregateEvent,
         string message = null, IMetadata metadata = null)
        where TEvent : IAggregateEvent<TAggregate, TIdentity>
    {
        Emit(aggregateEvent, metadata);
        throw DomainError.With(message ?? $"an unexpected error occured of type {typeof (TEvent).FullName}");
    }

    protected TAggregateState State { get; }
}

public abstract class BaseAggregateState<TAggregate, TIdentity, TEventApplier>
    : AggregateState<TAggregate, TIdentity, TEventApplier>,
      IAggregateState<TAggregate>
    where TEventApplier : class, IEventApplier<TAggregate, TIdentity>
    where TAggregate : IAggregateRoot<TIdentity>
    where TIdentity : IIdentity
{
    protected BaseAggregateState(TIdentity id)
    {
        Id = id;
    }

    public TIdentity Id { get; }
}


public abstract class BaseAggregateState<TAggregate, TIdentity, TEntity, TEventApplier>
    : BaseAggregateState<TAggregate, TIdentity, TEventApplier>,
      IAggregateState<TAggregate, TEntity>
    where TEventApplier : class, IEventApplier<TAggregate, TIdentity>
    where TAggregate : IAggregateRoot<TIdentity>
    where TIdentity : IIdentity
    where TEntity : class, IEntity
{
    protected BaseAggregateState(TIdentity id)
        : base(id)
    {
        Root = (TEntity)Activator.CreateInstance(typeof(TEntity), id);
    }

    public TEntity Root { get; protected set; }
}


public interface IAggregateState
{
}

public interface IAggregateState<TAggregateRoot> : IAggregateState
    where TAggregateRoot : IAggregateRoot
{
}

public interface IAggregateState<TAggregateRoot, out TEntity> : IAggregateState
    where TAggregateRoot : IAggregateRoot
    where TEntity : IEntity
{
    TEntity Root { get; }
}


```
